Refactor marrmot recipe and diagnostic to use ERA5 daily data made by new cmorizer

SarahAlidoost opened the pull request:
Before you start, read [CONTRIBUTING.md](https://github.com/ESMValGroup/ESMValTool/blob/master/CONTRIBUTING.md) and the [guide for diagnostic developers](https://esmvaltool.readthedocs.io/en/latest/esmvaldiag/index.html).

Please discuss your idea with the development team before getting started, to avoid disappointment later. The way to do this is to open a new issue on GitHub. If you are planning to modify an existing functionality, please discuss it with the original author(s) by tagging them in the issue.

* * *

**Tasks**

-   ~~[ ] [Create an issue](https://github.com/ESMValGroup/ESMValTool/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)~~
-   [x] Give this pull request a descriptive title that can be used as a one line summary in a changelog
-   [x] Make sure your code is composed of functions of no more than 50 lines and uses meaningful names for variables
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Preferably Codacy code quality checks pass, however a few remaining hard to solve Codacy issues are still acceptable. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes 
-   ~~[ ] (Only if really necessary) Add any additional dependencies needed for the diagnostic script to setup.py, esmvaltool/install/R/r_requirements.txt or esmvaltool/install/Julia/julia_requirements.txt (depending on the language of your script) and also to meta.yaml for conda dependencies (includes Python and others, but not R/Julia)~~
-   ~~[ ] If new dependencies are introduced, check that the license is compatible with [Apache2.0](https://github.com/ESMValGroup/ESMValTool/blob/master/LICENSE)~~

New recipe/diagnostic

-   ~~[ ] Add documentation for the recipe to the `doc/sphinx/source/recipes` folder and add a new entry to index.rst~~
-   ~~[ ] Add provenance information~~

Modified recipe/diagnostic

-   [x] Update documentation for the recipe to the `doc/sphinx/source/recipes` folder
-   ~~[ ] Update provenance information if needed~~
-   [x] Assign the author(s) of the affected recipe(s) as reviewer(s)

~~New data reformatting script~~

-   [ ] Test the CMORized data using recipes/example/recipe_check_obs.yml, to make sure the CMOR checks pass without errors
-   [ ] Add the new dataset to the table in the [documentation](https://esmvaltool.readthedocs.io/en/latest/getting_started/inputdata.html)
-   [ ] Tag @mattiarighi in this pull request, so that the new dataset can be added to the OBS data pool at DKRZ and synchronized with CEDA-Jasmin

~~Modified data reformatting script~~

-   [ ] Test the CMORized data using recipes/example/recipe_check_obs.yml, to make sure the CMOR checks pass without errors
-   [ ] Tag @mattiarighi in this pull request, so that the updated dataset can be added to the OBS data pool at DKRZ and synchronized with CEDA-Jasmin

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes {Link to corresponding issue}

SarahAlidoost commented:
For now, this PR is blocked on #1599

Peter9192 commented on esmvaltool/diag_scripts/hydrology/marrmot.py:
```diff
@@ -79,6 +79,20 @@ def _get_extra_info(cube):
     return time_start_end, lat_lon
 
 
+def fix_era5_time_coordinate(cube):
```
```suggestion
def shift_era5_time_coordinate(cube):
```
Because actually it's not a fix, but rather a workaround to make the variables with different time points compatible.

Peter9192 commented on esmvaltool/diag_scripts/hydrology/marrmot.py:
```diff
@@ -79,6 +79,20 @@ def _get_extra_info(cube):
     return time_start_end, lat_lon
 
 
+def fix_era5_time_coordinate(cube):
```
```suggestion
def fix_era5_time_coordinate(cube, shift=30):
```
Not strictly necessary, but perhaps it could be nice for future use cases?

SarahAlidoost replied:
@Peter9192 thanks. I didn't to apply this suggestion because I think the shift value (here 30min) is not something that changes in the future.

Peter9192 commented on esmvaltool/diag_scripts/hydrology/marrmot.py:
```diff
@@ -79,6 +79,20 @@ def _get_extra_info(cube):
+    For aggregated variables, already time format is [1990, 1, 1, 12, 0, 0].
+    """
+    time = cube.coord(axis='T')
+    time.points = time.points + 1 / 48
```
```suggestion
    time.points = time.points + shift / (24 * 60)
```
This will enable an arbitrary value of `shift` in minutes.

Peter9192 commented on esmvaltool/diag_scripts/hydrology/marrmot.py:
```diff
@@ -93,6 +107,11 @@ def main(cfg):
+        # Fix time coordinate of ERA5 instantaneous variables
+        if dataset == 'ERA5':
+            fix_era5_time_coordinate(all_vars['psl'])
+            fix_era5_time_coordinate(all_vars['tas'])
```
```suggestion
            shift_era5_time_coordinate(all_vars['tas'])
```

Peter9192 replied:
```suggestion
            shift_era5_time_coordinate(all_vars['tas'])
```

Peter9192 commented on esmvaltool/recipes/hydrology/recipe_marrmot.yml:
```diff
@@ -16,19 +16,16 @@ preprocessors:
     extract_shape:
       # Lumped model: needs catchment-aggregated input data
-      shapefile: meuse_hydrosheds.shp  # In aux (config-user.yml)
+      shapefile: Meuse/meuse.shp  # In aux (config-user.yml)
```
Note that with the latest version we're using, I think we now use a capital for Meuse.shp

Peter9192 commented on esmvaltool/recipes/hydrology/recipe_marrmot.yml:
```diff
@@ -45,24 +42,4 @@ diagnostics:
-      script:
-        script: hydrology/marrmot.py
-        basin: Meuse
+        basin: meuse
```
```suggestion
        basin: Meuse
```
Previous comment may also affect this one.

SarahAlidoost commented:
> Nice work! Perhaps we should consider moving this new function to `derive_evspsblpot.py`. Since that's where the two variables need to be aligned in order to calculate PET. If I want to use the same function for wflow, it would make more sense to import it from there than from the marrmot diagnostic.

@Peter9192 Thanks. Please note that the shift is only applied to era5 data and not era-interim. Also, this shift is considered as additional data processing. So, I don't see the `derive_evspsblpot.py` is a proper place for that. The diagnostic script is a better place. Also, I suggest copying this function to wflow diagnostic rather than importing. This avoid making diagnostics dependent on each other. This is in line with what you mentioned.

Peter9192 commented:
> So, I don't see the `derive_evspsblpot.py` is a proper place for that.

I agree that maybe this is also not the best place, but I also anticipate that we'll be seeing this issue more often in the future, and I think we should avoid duplicating code to fix it. I'm okay with this right now, but let's think about a better central place.

bouweandela commented:
@SarahAlidoost Could you please give the pull request a meaningful title, that can be used in the change log? Also, it would be great if you could use the check boxes in the pull request template, i.e. remove/cross out those that are not applicable and check those that have been done.

SarahAlidoost commented:
> @SarahAlidoost Could you please give the pull request a meaningful title, that can be used in the change log? Also, it would be great if you could use the check boxes in the pull request template, i.e. remove/cross out those that are not applicable and check those that have been done.

@bouweandela Thanks. I addressed your comment.
